File: benchmarks/fixtures/webapp_ts/src/models/payment.ts
/** Payment model. */
import { PaymentStatus } from './types';
import { getLogger } from '../utils/helpers';

const logger = getLogger("models.payment");

export interface PaymentRecord {
    id: string;
    userId: string;
    amount: number;
    currency: string;
    transactionId: string;
    status: PaymentStatus;
    createdAt: number;
    completedAt: number | null;
}

export class Payment {
    public readonly id: string;
    public readonly userId: string;
    public readonly amount: number;
    public readonly currency: string;
    public readonly transactionId: string;
    public status: PaymentStatus;
    public readonly createdAt: number;
    public completedAt: number | null;

    constructor(data: PaymentRecord) {
        this.id = data.id;
        this.userId = data.userId;
        this.amount = data.amount;
        this.currency = data.currency;
        this.transactionId = data.transactionId;
        this.status = data.status;
        this.createdAt = data.createdAt;
        this.completedAt = data.completedAt;
    }

    complete(): void {
        this.status = PaymentStatus.Completed;
        this.completedAt = Date.now();
        logger.info(`Payment ${this.transactionId} completed`);
    }

    fail(reason: string): void {
        this.status = PaymentStatus.Failed;
        logger.info(`Payment ${this.transactionId} failed: ${reason}`);
    }

    refund(): void {
        this.status = PaymentStatus.Refunded;
        logger.info(`Payment ${this.transactionId} refunded`);
    }

    isCompleted(): boolean {
        return this.status === PaymentStatus.Completed;
    }

    toJSON(): Record<string, unknown> {
        return {
            id: this.id,
            userId: this.userId,
            amount: this.amount,
            currency: this.currency,
            transactionId: this.transactionId,
            status: this.status,
            createdAt: this.createdAt,
            completedAt: this.completedAt,
        };
    }
}

File: benchmarks/fixtures/webapp_rb/models/payment.rb
# Payment model.

require_relative '../utils/helpers'

# Payment statuses.
module PaymentStatus
  PENDING = 'pending'
  PROCESSING = 'processing'
  COMPLETED = 'completed'
  FAILED = 'failed'
  REFUNDED = 'refunded'
end

# Payment record.
class Payment
  attr_reader :id, :user_id, :amount, :currency, :transaction_id, :created_at
  attr_accessor :status, :completed_at

  def initialize(id:, user_id:, amount:, currency:, transaction_id:, status: PaymentStatus::PENDING, created_at: nil, completed_at: nil)
    @id = id
    @user_id = user_id
    @amount = amount
    @currency = currency
    @transaction_id = transaction_id
    @status = status
    @created_at = created_at || Time.now.to_i
    @completed_at = completed_at
  end

  def complete
    @status = PaymentStatus::COMPLETED
    @completed_at = Time.now.to_i
    get_logger('models.payment').info("Payment #{@transaction_id} completed")
  end

  def fail(reason)
    @status = PaymentStatus::FAILED
    get_logger('models.payment').info("Payment #{@transaction_id} failed: #{reason}")
  end

  def do_refund
    @status = PaymentStatus::REFUNDED
    get_logger('models.payment').info("Payment #{@transaction_id} refunded")
  end

  def completed?
    @status == PaymentStatus::COMPLETED
  end

  def to_h
    {
      id: @id,
      user_id: @user_id,
      amount: @amount,
      currency: @currency,
      transaction_id: @transaction_id,
      status: @status,
      created_at: @created_at,
      completed_at: @completed_at
    }
  end
end
